semana-17/Food4U/src/business/usecases/Authentication/Login.ts

```diff
@@ -0,0 +1,40 @@
+import { AuthenticationGateway } from './../../gateways/authentication/authenticationGateway';
+import { CryptographyGateway } from './../../gateways/crypt/cryptographyGateway';
+import { UserGateway } from '../../gateways/user/UserGateway';
+
+export class LoginUseCase {
+    constructor(
+        private userGateway: UserGateway,
+        private cryptographyGateway: CryptographyGateway,
+        private authenticationGateway: AuthenticationGateway
+    ) {}
+
+    async execute(input: LoginInput): Promise<LoginOutPut> {
+        const user = await this.userGateway.getUserByEmail(input.email);
+        const passwordAuthentication = await this.cryptographyGateway.compare(
+            input.password,
+            user.getPassword()
+        );
+
+        if (!passwordAuthentication) {
+            throw new Error("Email or password are invalid");
```

Comment:
Me parece que pra cair nesse erro, o problema é necessariamente a senha errada, nao?

Reply:
eu coloquei assim pensando em não deixar claro pro usuário que é o password que tá errado. Assim, se for alguém tentando hackear, vai desmotiva-lo a ficar tentando várias senhas até acertar.